[BUG] Android project won't compile with Analysis

**Describe the bug**

When added to an android project compilation fails with

```
java.lang.AssertionError: Duplicated JavaClassDescriptor <redacted>/BuildConfig reported to IC
	at org.jetbrains.kotlin.incremental.JavaClassesTrackerImpl.onCompletedAnalysis(JavaClassesTrackerImpl.kt:64)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM$analyzeFilesWithJavaIntegration$2.invoke(TopDownAnalyzerFacadeForJVM.kt:110)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration(TopDownAnalyzerFacadeForJVM.kt:122)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration$default(TopDownAnalyzerFacadeForJVM.kt:96)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:262)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:53)
	at org.jetbrains.kotlin.cli.common.messages.AnalyzerWithCompilerReport.analyzeAndReport(AnalyzerWithCompilerReport.kt:113)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.analyze(KotlinToJVMBytecodeCompiler.kt:253)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.repeatAnalysisIfNeeded(KotlinToJVMBytecodeCompiler.kt:223)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.repeatAnalysisIfNeeded(KotlinToJVMBytecodeCompiler.kt:223)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:100)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli$default(KotlinToJVMBytecodeCompiler.kt:58)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:170)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:52)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:92)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:44)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:98)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:434)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:120)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally(IncrementalCompilerRunner.kt:357)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally$default(IncrementalCompilerRunner.kt:299)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl$rebuild(IncrementalCompilerRunner.kt:118)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl(IncrementalCompilerRunner.kt:169)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile(IncrementalCompilerRunner.kt:80)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.execIncrementalCompiler(CompileServiceImpl.kt:622)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.access$execIncrementalCompiler(CompileServiceImpl.kt:100)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1713)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:359)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:562)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:796)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:677)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:676)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**To Reproduce**
Steps to reproduce the behavior:
1. Create new android project.
2. Add Analysis plugin

**Expected behavior**
Project should compile


**Environment (please complete the following information):**
 - Kotlin version [eg. 1.3.61]: 1.6.0 
 - [Kotlin Intellij IDEA plugin version](https://plugins.jetbrains.com/plugin/6954-kotlin/versions) [eg. 1.3.61-release-IJ2019.3-1]: 211-1.6.10-release-923-AS7442.40
- Intellij IDEA version:  AS 2021.1.1
- Gradle version: 7.2

--- a/docs/docs/analysis/README.md
+++ b/docs/docs/analysis/README.md
@@ -60,6 +60,20 @@ apply plugin: 'io.arrow-kt.analysis.kotlin'
 
 This adds both the Kotlin compiler plug-in -- which performs the checks -- and the pre and post-conditions for the Kotlin standard library. You are ready to get your first analysis results.
 
+### Λrrow Analysis + Android
+
+If you want to use the plug-in in an Android project, you may run into [this Kotlin compiler issue](https://youtrack.jetbrains.com/issue/KT-38576), characterized by the following error message:
+
+```
+java.lang.AssertionError: Duplicated JavaClassDescriptor ... reported to IC
+```
+
+To solve this problem you have to disable precise Java tracking, by adding the following line in your `gradle.properties` file:
+
+```kotlin
+kotlin.incremental.usePreciseJavaTracking=false
+```
+
 ## Running the analysis
 
 Open a new file and write the following line. This code is incorrect because you want to obtain the third element of an empty list.

--- a/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/ast/KotlinInterpreter.kt
+++ b/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/ast/KotlinInterpreter.kt
@@ -6,7 +6,6 @@ import arrow.meta.plugins.analysis.phases.analysis.solver.ast.context.elements.E
 import arrow.meta.plugins.analysis.phases.analysis.solver.ast.kotlin.descriptors.*
 import arrow.meta.plugins.analysis.phases.analysis.solver.ast.kotlin.elements.*
 import org.jetbrains.kotlin.descriptors.*
-import org.jetbrains.kotlin.descriptors.impl.LazyClassReceiverParameterDescriptor
 import org.jetbrains.kotlin.descriptors.impl.LocalVariableDescriptor
 import org.jetbrains.kotlin.descriptors.impl.TypeAliasConstructorDescriptor
 import org.jetbrains.kotlin.psi.*
@@ -14,14 +13,11 @@ import org.jetbrains.kotlin.psi.psiUtil.isNull
 
 /* ktlint-enable no-wildcard-imports */
 
-private fun <
-  A : org.jetbrains.kotlin.descriptors.DeclarationDescriptor, B : DeclarationDescriptor> A.repr():
-  B = this as B
-
+@Suppress("UNUSED_PARAMETER", "UNCHECKED_CAST")
 private fun <A : DeclarationDescriptor, B : DeclarationDescriptor> A.repr(unit: Unit = Unit): B =
   this as B
 
-fun <A : Element, B : Element> A.repr(): B = this as B
+@Suppress("UNCHECKED_CAST") fun <A : Element, B : Element> A.repr(): B = this as B
 
 fun <
   A : org.jetbrains.kotlin.descriptors.DeclarationDescriptor, B : DeclarationDescriptor> A.model():
@@ -36,17 +32,18 @@ fun <
     is ClassDescriptor -> KotlinClassDescriptor(this).repr()
     is TypeAliasDescriptor -> KotlinTypeAliasDescriptor(this).repr()
     is ValueParameterDescriptor -> KotlinValueParameterDescriptor(this).repr()
-    // is ModuleDescriptor -> KotlinModuleDescriptor(this).repr()
-    is LazyClassReceiverParameterDescriptor -> KotlinReceiverParameterDescriptor(this).repr()
+    is ModuleDescriptor -> KotlinModuleDescriptor(this).repr()
     is ReceiverParameterDescriptor -> KotlinReceiverParameterDescriptor(this).repr()
     is LocalVariableDescriptor -> KotlinLocalVariableDescriptor(this).repr()
     is PackageFragmentDescriptor -> KotlinPackageFragmentDescriptor(this).repr()
     // fallback cases: we sometimes find unknown descriptors
     // and for those cases we need nothing else than what the
     // abstract classes provide
-    is FunctionDescriptor -> (object : KotlinFunctionDescriptor(this) {}).repr()
-    is VariableDescriptor -> (object : KotlinVariableDescriptor(this) {}).repr()
-    else -> TODO("Missing impl for $this (${this.javaClass.name})")
+    is FunctionDescriptor -> KotlinDefaultFunctionDescriptor(this).repr()
+    is VariableDescriptor -> KotlinDefaultVariableDescriptor(this).repr()
+    is DeclarationDescriptorWithVisibility ->
+      KotlinDefaultDeclarationDescriptorWithVisibility(this).repr()
+    else -> KotlinDefaultDeclarationDescriptor(this).repr()
   }
 
 fun <A : KtElement, B : Element> A.model(): B =
@@ -152,4 +149,4 @@ fun <A : KtElement, B : Element> A.model(): B =
       )
   }
 
-fun <A : Element, B : KtElement> A.element(): B = impl() as B
+@Suppress("UNCHECKED_CAST") fun <A : Element, B : KtElement> A.element(): B = impl() as B

--- a/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinDeclarationDescriptor.kt
+++ b/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinDeclarationDescriptor.kt
@@ -30,3 +30,9 @@ fun interface KotlinDeclarationDescriptor : DeclarationDescriptor {
   override val name: Name
     get() = Name(impl().name.asString())
 }
+
+class KotlinDefaultDeclarationDescriptor(
+  val impl: org.jetbrains.kotlin.descriptors.DeclarationDescriptor
+) : KotlinDeclarationDescriptor {
+  override fun impl(): org.jetbrains.kotlin.descriptors.DeclarationDescriptor = impl
+}

--- a/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinDeclarationDescriptorWithVisibility.kt
+++ b/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinDeclarationDescriptorWithVisibility.kt
@@ -9,3 +9,9 @@ fun interface KotlinDeclarationDescriptorWithVisibility :
   override val visibility: Visibility
     get() = KotlinVisibility { impl().visibility.delegate }
 }
+
+class KotlinDefaultDeclarationDescriptorWithVisibility(
+  val impl: org.jetbrains.kotlin.descriptors.DeclarationDescriptorWithVisibility
+) : KotlinDeclarationDescriptorWithVisibility {
+  override fun impl(): org.jetbrains.kotlin.descriptors.DeclarationDescriptorWithVisibility = impl
+}

--- a/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinFunctionDescriptor.kt
+++ b/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinFunctionDescriptor.kt
@@ -29,3 +29,6 @@ abstract class KotlinFunctionDescriptor(
   override val isSuspend: Boolean
     get() = impl().isSuspend
 }
+
+class KotlinDefaultFunctionDescriptor(impl: org.jetbrains.kotlin.descriptors.FunctionDescriptor) :
+  KotlinFunctionDescriptor(impl)

--- a/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinModuleDescriptor.kt
+++ b/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinModuleDescriptor.kt
@@ -11,7 +11,7 @@ class KotlinModuleDescriptor(val impl: org.jetbrains.kotlin.descriptors.ModuleDe
 
   override fun impl(): org.jetbrains.kotlin.descriptors.ModuleDescriptor = impl
 
-  override fun getPackage(pck: String): PackageViewDescriptor? =
+  override fun getPackage(pck: String): PackageViewDescriptor =
     impl().getPackage(org.jetbrains.kotlin.name.FqName(pck)).model()
 
   override fun getSubPackagesOf(fqName: FqName): List<FqName> =

--- a/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinVariableDescriptor.kt
+++ b/plugins/analysis/kotlin-plugin/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/ast/kotlin/descriptors/KotlinVariableDescriptor.kt
@@ -15,3 +15,6 @@ abstract class KotlinVariableDescriptor(
   override val isLateInit: Boolean
     get() = impl().isLateInit
 }
+
+class KotlinDefaultVariableDescriptor(impl: org.jetbrains.kotlin.descriptors.VariableDescriptor) :
+  KotlinVariableDescriptor(impl)

